pipeline/merger/src/main/scala/uk/ac/wellcome/platform/merger/rules/FieldMergeRule.scala

```diff
@@ -26,30 +29,36 @@ import uk.ac.wellcome.platform.merger.rules.WorkPredicates.WorkPredicate
 trait FieldMergeRule {
   protected final type Params = (UnidentifiedWork, Seq[TransformedBaseWork])
   protected type FieldData
+  protected type MergeState = State[Set[TransformedBaseWork], FieldData]
+
+  def apply(targets: UnidentifiedWork,
+            sources: Seq[TransformedBaseWork]): MergeState =
+    merge(targets, sources) match {
+      case FieldMergeResult(field, ruleRedirects) =>
+        State(existingRedirects =>
+          (existingRedirects ++ ruleRedirects.toSet, field))
+    }
 
   def merge(target: UnidentifiedWork,
             sources: Seq[TransformedBaseWork]): FieldMergeResult[FieldData]
 
-  protected val identityOnTarget: PartialFunction[Params, UnidentifiedWork] = {
-    case (target, _) => target
-  }
-
-  protected trait PartialRule extends PartialFunction[Params, FieldData] {
+  protected trait PartialRule {
     val isDefinedForTarget: WorkPredicate
     val isDefinedForSource: WorkPredicate
 
-    def rule(target: UnidentifiedWork,
-             sources: Seq[TransformedBaseWork]): FieldData
+    protected def rule(target: UnidentifiedWork,
+                       sources: NonEmptyList[TransformedBaseWork]): FieldData
```

Comment:
Is there not a way to update all the other sources to use this instead of proliferating how we apply this trait, which is a little complicated as is.

Reply:
I believe this PR already does that? It's equivalent or slightly simpler to what we had before, except now we compose data rather than operations.

Reply:
Could we remove the other `apply`s then?

Reply:
There are only 2 (`PartialRule` is a nested trait within `FieldMergeRule`) and one of them is just a convenience to avoid having to construct single-item lists when calling the `PartialRule`, it doesn't contain any logic